### ISSUE TITLE
No guarantee about the input bound 

### ISSUE THREAD

**Issue by renqibing** [bethgelab/foolbox#602]
**In attacks/base.py, inside the function '__call__' with keyword 'final'**, it should first check if input 'x' exceeds the 'model.bounds' or not. Because if 'x' does not reside in the bound, when it is sent into self.distance.clip_perturbation() line, the difference between **x and xp** is not equal to perturbation, which results in the failure of attack methods.

**Comment by jonasrauber** [comment]
Hmm, I think I understand what you are saying. And we could probably add a check for that. But I am not sure it's necessary. You pass in some clean, unperturbed data (e.g., images from the dataset). They should, by definition, be within the bounds.

Of course, checking for this could help the user to avoid certain mistakes. But there are many ways to use something in the wrong way, we will not be able to catch them all and this does not seem like a problem one would commonly run into.

Could you explain why you ran into it / how often that happens.

**Comment by renqibing** [comment]
Hi,

I tried to use your tools to generate adversarial examples for training. Before that, I normalized CIFAR10 training dataset and then the scale of it is beyond (0, 1). In such case, when performing some kind of fixed-epsilon attack(PGD, FGSM) on data, the perturbation does not work since some examples already exceed (0, 1) by larger value than epsilon. You can check the code “xpc = self.distance.clip_perturbation(x, xp, epsilon)” in attacks/base.py. 

Thanks,
Qibing
发件人: Jonas Rauber
发送时间: 2020年12月18日 23:44
收件人: bethgelab/foolbox
抄送: Qibing Ren; Author
主题: Re: [bethgelab/foolbox] No guarantee about the input bound (#602)

Hmm, I think I understand what you are saying. And we could probably add a check for that. But I am not sure it's necessary. You pass in some clean, unperturbed data (e.g., images from the dataset). They should, by definition, be within the bounds.
Of course, checking for this could help the user to avoid certain mistakes. But there are many ways to use something in the wrong way, we will not be able to catch them all and this does not seem like a problem one would commonly run into.
Could you explain why you ran into it / how often that happens.

**Comment by jonasrauber** [comment]
I think you are using it wrong. The bounds and the inputs need to match. You can let Foolbox handle the normalization using the preprocessing parameter.

**Comment by renqibing** [comment]
I see. Thank you for your response. It’s my wrong usage that causes this problem. Maybe is it better to throw a warning when some users, like me, try to send inputs that do not match the model bounds? 
发件人: Jonas Rauber
发送时间: 2020年12月20日 16:40
收件人: bethgelab/foolbox
抄送: Qibing Ren; Author
主题: Re: [bethgelab/foolbox] No guarantee about the input bound (#602)

I think you are using it wrong. The bounds and the inputs need to match. You can let Foolbox handle the normalization using the preprocessing parameter.

**Comment by jonasrauber** [comment]
Yes, maybe we should catch this problem.

**Comment by mipet23** [comment]
I use Cifar10 data to train my neural network. To attack my network with PGD I pass my preprocessing and set the bounds to (0,1), but the success of the attack stays very low. 

`preprocessing = dict(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], axis=-3)
fb_model = fbox.PyTorchModel(model, bounds=(0, 1), preprocessing=preprocessing)`

If I train on non-normalized data and perform the attack, the attack is very successful. What am I doing wrong? I would greatly appreciate your help.

**Comment by Changgun-Choi** [comment]
Hello, is there answer for the question above? i couldn't find it so!
Thank you very much
